akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala

```diff
@@ -36,240 +35,135 @@ private object PoolSlot {
-          }) :: Nil
-        } else {
-          inflightRequests.map { rc ⇒
-            if (rc.retriesLeft == 0) {
-              val reason = error.fold[Throwable](new UnexpectedDisconnectException("Unexpected disconnect"))(ConstantFun.scalaIdentityFunction)
-              connInport ! ActorPublisherMessage.Cancel
-              ResponseDelivery(ResponseContext(rc, Failure(reason)))
-            } else SlotEvent.RetryRequest(rc.copy(retriesLeft = rc.retriesLeft - 1))
-          }(collection.breakOut)
-        }
+        override def onUpstreamFailure(ex: Throwable): Unit = disconnect(Some(ex))
       }
-      inflightRequests = immutable.Queue.empty
-      onNext(SlotEvent.Disconnected(slotIx, results.size) :: results)
-      if (canceled) onComplete()
 
-      context.become(unconnected)
-    }
+      // upstream pushes we create the inner stream if necessary or push if we're already connected
+      override def onPush(): Unit = {
+        def establishConnectionFlow() = {
+          connectionFlowSource = new SubSourceOutlet[HttpRequest]("RequestSource")
+          connectionFlowSource.setHandler(connectionOutFlowHandler)
 
-    override def onComplete(): Unit = {
-      exposedPublisher.shutdown(None)
-      super.onComplete()
-      shutdown()
-    }
+          connectionFlowSink = new SubSinkInlet[HttpResponse]("ResponseSink")
+          connectionFlowSink.setHandler(connectionInFlowHandler)
 
-    override def onError(cause: Throwable): Unit = {
-      exposedPublisher.shutdown(Some(cause))
-      super.onError(cause)
-      shutdown()
-    }
+          isConnected = true
 
-    def shutdown(): Unit = context.stop(self)
-  }
+          Source.fromGraph(connectionFlowSource.source)
+            .via(connectionFlow).runWith(Sink.fromGraph(connectionFlowSink.sink))(subFusingMaterializer)
 
-  private case class FromConnection(ev: Any) extends NoSerializationVerificationNeeded
-
-  private class FlowInportActor(slotProcessor: ActorRef) extends ActorPublisher[HttpRequest] with ActorLogging {
-    def receive: Receive = {
-      case ev: Request            ⇒ slotProcessor ! FromConnection(ev)
-      case OnNext(r: HttpRequest) ⇒ onNext(r)
-      case OnComplete             ⇒ onCompleteThenStop()
-      case OnError(e)             ⇒ onErrorThenStop(e)
-      case Cancel ⇒
-        slotProcessor ! FromConnection(Cancel)
-        context.stop(self)
-    }
-  }
-  def flowInportProps(s: ActorRef) = Props(new FlowInportActor(s)).withDeploy(Deploy.local)
-
-  private class FlowOutportActor(slotProcessor: ActorRef) extends ActorSubscriber with ActorLogging {
-    def requestStrategy = ZeroRequestStrategy
-    def receive: Receive = {
-      case Request(n) ⇒ request(n)
-      case Cancel     ⇒ cancel()
-      case ev: OnNext ⇒ slotProcessor ! FromConnection(ev)
-      case ev @ (OnComplete | OnError(_)) ⇒
-        slotProcessor ! FromConnection(ev)
-        context.stop(self)
+          connectionFlowSink.pull()
```

Comment:
I don't think we want to pull here, I think that should come when there is demand at the connectionFlowSource but I wasn't able to get that to work.